File: Trie/Java/Trie.java
import java.io.*;
import java.util.Scanner;

class Trie{

    Trienode root;

    public Trie(){
        this.root = null;
    }

    public static Trienode initTrienode(Trienode r,int s){

        //System.out.println("test");
        r = new Trienode();
        r.tn = new THashtable(s);

        for(int i=0;i<s;i++){
            //System.out.println("test");
            r.tn.slots[i] = new TNode(null);
        }
        

        return r;
        
    }

    public static void resizeTrienode(Trienode r,int new_size){
        THashtable t = new THashtable(new_size);

        for(int i=0;i<new_size;i++){
            if(i < r.tn.size)
            t.slots[i] = r.tn.slots[i];
            else t.slots[i] = new TNode(null);
        }

        t.num_c = r.tn.num_c;
        r.tn = t;
    }

    public static Trienode insertString(Trienode r,String w){

        TNode aux = null;

        for(int i=0;i<w.length();i++){
            int index = w.charAt(i) -'a';

            if(i == 0){

                if(r == null)
                r = initTrienode(r,index+1);

                if(index >= r.tn.size)
                resizeTrienode(r,index+1);

                if(r.tn.slots[index].key == null){
                r.tn.slots[index].key = w.charAt(i) + "";
                r.tn.num_c++;
                }

                aux = r.tn.slots[index];
            }else{
                if(aux.child == null)
                    aux.child = initTrienode(aux.child,index+1);
                    
               
                if(index >= aux.child.tn.size)
                resizeTrienode(aux.child,index+1);
                
                if(aux.child.tn.slots[index].key == null){
                    aux.child.tn.slots[index].key = w.charAt(i)+"";
                    aux.child.tn.num_c++;
                }
                   
                aux = aux.child.tn.slots[index];    
             }

            }

        return r;
    }

    public static void printTrie(Trienode r,String k,int ct){

        if( r!= null){
            for(int i=0;i<r.tn.size;i++){

                if(r.tn.slots[i].key != null){
                    k = k + r.tn.slots[i].key;
                    System.out.print(k+" | ");
                    ct++;
                    printTrie(r.tn.slots[i].child,k,ct);
                    k = k.substring(0,k.length() - 1);
                    ct--;

                    if(ct == 0)
                    System.out.println("");


                }
                
            }
        }
    }

    public static void main(String args[]){

        Scanner sc = new Scanner(System.in);

        Trie t = new Trie();

        while(true){
            System.out.println("Give me a word ");
            String w = sc.next();
            t.root = insertString(t.root,w);
            System.out.println(t.root.tn.size);
            printTrie(t.root,"",0);
            System.out.println("");
        }
        

        

    }
}

File: Trie/Python/Trie.py
import os
import sys


from TNode import TNode
from THashtable import THashtable
from Trienode import Trienode

class Trie:

    def __init__(self):
        self.root = None

    def initTrienode(self,r:Trienode,init_size:int):
        r = Trienode()
        r.tn = THashtable(init_size)

        for i in range(init_size):
            r.tn.slots[i] = TNode(None)
        
        return r

    def resizeTrienode(self,r:Trienode,new_size:int):
        t = THashtable(new_size)
        t.size = new_size

        for i in range(new_size):
            if i < r.tn.size:
                t.slots[i] = r.tn.slots[i]
            else:
                t.slots[i] = TNode(None)
        
        t.num_c = r.tn.num_c
        r.tn = t
    
    def insertString(self,r:Trienode,w):
        aux = TNode(None)
        
        for i in range(len(w)):
            index = ord(w[i]) - ord('a')
        
            if i == 0:
                if r == None:
                    r = self.initTrienode(r,index+1)
                
                if index >= r.tn.size:
                    self.resizeTrienode(r,index+1)
                
                if r.tn.slots[index].key == None:
                    r.tn.slots[index].key = w[i]
                    r.tn.num_c = r.tn.num_c + 1
                   
                aux = r.tn.slots[index]   
            else:
                if aux.child == None:
                    aux.child = self.initTrienode(aux.child,index+1)
                
                if index >= aux.child.tn.size:
                    resizeTrienode(aux.child,index+1)

                if aux.child.tn.slots[index].key == None:
                    aux.child.tn.slots[index].key = w[i]
                    aux.child.tn.num_c = aux.child.tn.num_c + 1

                aux = aux.child.tn.slots[index]

        return r


    def printTrienode(self,r:Trienode,k,ct):
    
        if r != None:
            i = 0
            j = 0
            b = False
           
            while i < r.tn.size and not(b):
                if j >= r.tn.num_c:
                    b = True
                else:
                    if r.tn.slots[i].key != None:
                        j = j + 1
                        k = k + r.tn.slots[i].key
                        print(k,end=" | ")
                        ct = ct + 1
                        aux = r.tn.slots[i]
                        self.printTrienode(aux.child,k,ct)
                        k = k[:-1]
                        ct = ct - 1
                        
                        if ct == 0:
                            print("\n \n")
                
                i = i + 1
                        




"""Testing"""



t = Trie()


while True:
    print("Give me a word \n")
    wr = input()
    t.root = t.insertString(t.root,wr)
    t.printTrienode(t.root,"",0)